### PR TITLE
Add tests command

### LINKED ISSUE #7: Option to `flutter test`
VSCode has a visual option for running a single test:
flutter-test
I know that something visual is a bit complex, but what about a command like `flutter-run-all-tests` or `flutter-run-test-at-point`?

### PR DESCRIPTION
**Pull request by ericdallo**
Fix https://github.com/amake/flutter.el/issues/7

Adds 2 new commands:
* `flutter-test-all` - Run all tests from project root and display in a new buffer.
* `flutter-test-current-file` - Run tests from current file display in a new buffer.

This flutter test commands helps a lot, hope you like it too.

I'm thinking to add one more command `flutter-test-at-point`, running only the test on cursor, in other PR.

### CODE REVIEW

**Comment by ericdallo** [comment]
c/c @amake

File: flutter.el [hunk 1]
```diff
@@ -103,8 +104,9 @@ the `flutter` process."
        (error "Root of Flutter project not found"))))
 
 (defmacro flutter--with-run-proc (args &rest body)
-  "Execute BODY while ensuring an inferior `flutter` process is running.
+  "Execute BODY while ensuring an inferior `flutter run` process is running.
 
+BUFFER-NAME is the chosen buffer where will run the proc.
```
Comment (amake) [inline flutter.el]: There is no `BUFFER-NAME` parameter, so I don't understand this.

File: flutter.el [hunk 2]
```diff
@@ -118,17 +120,52 @@ ARGS is a space-delimited string of CLI flags passed to
...
+  `(flutter--from-project-root
+    (let* ((buffer (get-buffer-create flutter-test-buffer-name))
+           (alive (flutter--testing-p))
+           (arglist (when ,args (split-string ,args))))
+      (unless alive
+        (apply #'make-comint-in-buffer "Flutter Test" buffer (flutter-build-command) nil "test" arglist))
+      (with-current-buffer buffer
+        (unless (derived-mode-p 'flutter-mode)
+          (flutter-mode)))
+      ,@body)))
```
Comment (amake) [inline flutter.el]: This is a copy-and-paste of `flutter--with-run-proc`. Copy-and-paste is not bad in and of itself, but here I don't think it's appropriate:

`flutter--with-run-proc` is the way it is because `flutter run` is a long-running process that you want to send commands to. That's why there is all this machinery to make it easy to run things in an already extant buffer.

For test running, I expect it to be a one-shot execution that you do not care to communicate with. For this you don't need all of this machinery. I suggest you take a look at [`ruby-test-run` and `ruby-test-run-at-point` in ruby-test-mode](https://github.com/ruby-test-mode/ruby-test-mode/blob/e805a81c3ea787f3c82dbb252c21d747be20bc7a/ruby-test-mode.el#L359-L384).

**Review by amake — CHANGES_REQUESTED** [review]
Thanks for the PR.

I have been using equivalent functionality for Ruby (ruby-test-mode) to great effect lately, so I think this kind of thing could be a great addition.

File: flutter.el (hunk 1 above)
Comment (ericdallo) [reply flutter.el]: Sorry, I removed the code and forgot to undo this

File: flutter.el (hunk 2 above)
Comment (ericdallo) [reply flutter.el]: Nice! I'll fix it!

**Comment by ericdallo** [comment]
Done @amake!

File: flutter.el [hunk 3]
```diff
@@ -103,7 +103,7 @@ the `flutter` process."
        (error "Root of Flutter project not found"))))
 
 (defmacro flutter--with-run-proc (args &rest body)
-  "Execute BODY while ensuring an inferior `flutter` process is running.
+  "Execute BODY while ensuring an inferior `flutter run` process is running.
```
Comment (amake) [inline flutter.el]: Can you revert this change?

File: flutter.el [hunk 4]
```diff
@@ -119,16 +119,29 @@ ARGS is a space-delimited string of CLI flags passed to
...
 
 (defun flutter--running-p ()
-  "Return non-nil if the `flutter` process is already running."
+  "Return non-nil if the `flutter run` process is already running."
   (comint-check-proc flutter-buffer-name))
 
 (defun flutter--send-command (command)
   "Send COMMAND to a running Flutter process."
   (flutter--with-run-proc
+   flutter-buffer-name
```
Comment (amake) [inline flutter.el]: Can you revert these changes?

File: flutter.el [hunk 5]
```diff
@@ -119,16 +119,29 @@ ARGS is a space-delimited string of CLI flags passed to
...
+  "Execute `flutter test` inside Emacs.
+
+ARGS is a space-delimited string of CLI flags passed to
+`flutter`, and can be nil.  Call with a prefix to be prompted for
+args."
+  (interactive
+   (list (when current-prefix-arg
+           (read-string "Args: "))))
+  (flutter--from-project-root
+   (compilation-start (format "%s test %s" (flutter-build-command) args) t)))
```
Comment (amake) [inline flutter.el]: The function docstring says that `args` can be nil, but the resulting command string will be `flutter test nil`. I don't think that's what you want.

**Review by amake — CHANGES_REQUESTED** [review]

File: flutter.el (hunk 3 above)
Comment (ericdallo) [reply flutter.el]: Sorry, forgot about it

File: flutter.el (hunk 4 above)
Comment (ericdallo) [reply flutter.el]: Sorry, I forgot to remove this too.

File: flutter.el (hunk 5 above)
Comment (ericdallo) [reply flutter.el]: Thanks, I wrapped with an `or`

**Comment by ericdallo** [comment]
Fixed @amake :sweat_smile:

**Comment by amake** [comment]
Thanks for your contribution!
